Update the ut_control_keyStringMapping_t struct as getting build warning

update the ut_control_keyStringMapping_t 
from: typedef struct
      {
          char *string;
          int32_t value;
      } ut_control_keyStringMapping_t;
 
to: typedef struct
    {
        const char *string;
        int32_t value;
    } ut_control_keyStringMapping_t;

`/home/krt465/karthikeyan/Polaris_nonAV/hands-on/address_bhanu_RVC-3/rdk-halif-aidl-test-indicator/src/vts_indicator_helper_fu
nctions.cpp:68:6: warning: ISO C++ forbids converting a string constant to ‘char*’ [-Wwrite-strings]
   68 |     {"false", (int32_t) false},
      |      ^~~~~~~
/home/krt465/karthikeyan/Polaris_nonAV/hands-on/address_bhanu_RVC-3/rdk-halif-aidl-test-indicator/src/vts_indicator_helper_fu
nctions.cpp:69:6: warning: ISO C++ forbids converting a string constant to ‘char*’ [-Wwrite-strings]
   69 |     {"true",  (int32_t) true},
      |      ^~~~~~
/home/krt465/karthikeyan/Polaris_nonAV/hands-on/address_bhanu_RVC-3/rdk-halif-aidl-test-indicator/src/vts_indicator_helper_fu
nctions.cpp:75:6: warning: ISO C++ forbids converting a string constant to ‘char*’ [-Wwrite-strings]
   75 |     {"EX_NONE",                  (int32_t) Status::EX_NONE},
      |      ^~~~~~~~~
/home/krt465/karthikeyan/Polaris_nonAV/hands-on/address_bhanu_RVC-3/rdk-halif-aidl-test-indicator/src/vts_indicator_helper_fu
nctions.cpp:76:6: warning: ISO C++ forbids converting a string constant to ‘char*’ [-Wwrite-strings]
   76 |     {"EX_SECURITY",              (int32_t) Status::EX_SECURITY},`

--- a/include/ut_control_plane.h
+++ b/include/ut_control_plane.h
@@ -44,7 +44,7 @@ typedef enum
 
 typedef struct
 {
-    char *string;
+    const char *string;
     int32_t value;
 } ut_control_keyStringMapping_t;
 
@@ -160,7 +160,7 @@ uint32_t UT_Control_GetMapValue(const ut_control_keyStringMapping_t *conversionM
  * @param key The integer key for which the mapped string is being searched.
  * @return Pointer to the string associated with `key` if found, otherwise `NULL`.
  */
-char *UT_Control_GetMapString(const ut_control_keyStringMapping_t *conversionMap, int32_t key);
+const char *UT_Control_GetMapString(const ut_control_keyStringMapping_t *conversionMap, int32_t key);
 
 #ifdef __cplusplus
 }

--- a/src/ut_control_plane.c
+++ b/src/ut_control_plane.c
@@ -151,7 +151,7 @@ static void call_callback_on_match(cp_message_t *mssg, ut_cp_instance_internal_t
 
     pkvpInstance = ut_kvp_createInstance();
 
-    /* Note: mssg-message data will be freed by the destoryInstance() function */
+    /* Note: mssg->message will get freed after this callback */
     status = ut_kvp_openMemory(pkvpInstance, mssg->message, mssg->size );
     if (status != UT_KVP_STATUS_SUCCESS)
     {

--- a/src/ut_control_plane.c
+++ b/src/ut_control_plane.c
@@ -210,8 +210,6 @@ static void *service_state_machine(void *data)
             continue;
         }
 
-        bool free_message_payload = false;  // track if msg->message must be freed
-
         switch (msg->status)
         {
             case EXIT_REQUESTED:
@@ -225,7 +223,6 @@ static void *service_state_machine(void *data)
             {
                 UT_CONTROL_PLANE_DEBUG("DATA RECEIVED\n");
                 call_callback_on_match(msg, pInternal);
-                free_message_payload = true;
             }
             break;
 
@@ -236,16 +233,13 @@ static void *service_state_machine(void *data)
             }
         }
 
-        if (free_message_payload && msg->message != NULL)
-        {
-            free(msg->message);
-            msg->message = NULL;
-        }
-
         if (msg != NULL)
         {
+            if (msg->message != NULL)
+            {
+                free(msg->message);
+            }
             free(msg);
-            msg = NULL;
         }
     }
 

--- a/src/ut_control_plane.c
+++ b/src/ut_control_plane.c
@@ -193,7 +193,7 @@ static void *service_ws_requests(void *data)
 static void *service_state_machine(void *data)
 {
     ut_cp_instance_internal_t *pInternal = validateCPInstance((ut_controlPlane_instance_t*)data);
-    cp_message_t *msg;
+    cp_message_t *msg = NULL;
 
     if (pInternal == NULL)
     {
@@ -204,11 +204,13 @@ static void *service_state_machine(void *data)
 
     while (!pInternal->exit_request)
     {
-        msg = dequeue_message( pInternal );
+        msg = dequeue_message(pInternal);
         if (msg == NULL)
         {
             continue;
         }
+
+        bool free_message_payload = false;  // track if msg->message must be freed
 
         switch (msg->status)
         {
@@ -216,11 +218,6 @@ static void *service_state_machine(void *data)
             {
                 UT_CONTROL_PLANE_DEBUG("EXIT REQUESTED in thread1. Thread1 going to exit\n");
                 pInternal->exit_request = true;
-                if(msg)
-                {
-                    free(msg);
-                    msg = NULL;
-                }
             }
             break;
 
@@ -228,34 +225,37 @@ static void *service_state_machine(void *data)
             {
                 UT_CONTROL_PLANE_DEBUG("DATA RECEIVED\n");
                 call_callback_on_match(msg, pInternal);
-
-                free(msg->message);
-                msg->message = NULL;
-                free(msg);
-                msg = NULL;
+                free_message_payload = true;
             }
             break;
 
             default:
             {
                 UT_CONTROL_PLANE_ERROR("Unknown message type received: %d\n", msg->status);
-                if( msg)
-                {
-                    free(msg);
-                    msg = NULL;
-                }
+                break;
             }
-            break;
+        }
+
+        if (free_message_payload && msg->message != NULL)
+        {
+            free(msg->message);
+            msg->message = NULL;
+        }
+
+        if (msg != NULL)
+        {
+            free(msg);
+            msg = NULL;
         }
     }
 
     if (pthread_join(pInternal->ws_thread_handle, NULL) != 0)
     {
         UT_CONTROL_PLANE_ERROR("ws_thread_handle Failed to join from instance : %p \n", pInternal);
-        /*TODO: need to confirm if this return is required*/
         return NULL;
     }
-    UT_CONTROL_PLANE_DEBUG("Thread2 exitted\n");
+
+    UT_CONTROL_PLANE_DEBUG("Thread2 exited\n");
     return NULL;
 }
 

--- a/src/ut_control_plane.c
+++ b/src/ut_control_plane.c
@@ -562,7 +562,7 @@ uint32_t UT_Control_GetMapValue(const ut_control_keyStringMapping_t *conversionM
     return defaultValue;
 }
 
-char *UT_Control_GetMapString(const ut_control_keyStringMapping_t *conversionMap, int32_t key)
+const char *UT_Control_GetMapString(const ut_control_keyStringMapping_t *conversionMap, int32_t key)
 {
     if (conversionMap == NULL)
     {

--- a/src/ut_control_plane.c
+++ b/src/ut_control_plane.c
@@ -205,6 +205,10 @@ static void *service_state_machine(void *data)
     while (!pInternal->exit_request)
     {
         msg = dequeue_message( pInternal );
+        if (msg == NULL)
+        {
+            continue;
+        }
 
         switch (msg->status)
         {
@@ -212,6 +216,11 @@ static void *service_state_machine(void *data)
             {
                 UT_CONTROL_PLANE_DEBUG("EXIT REQUESTED in thread1. Thread1 going to exit\n");
                 pInternal->exit_request = true;
+                if(msg)
+                {
+                    free(msg);
+                    msg = NULL;
+                }
             }
             break;
 
@@ -219,13 +228,22 @@ static void *service_state_machine(void *data)
             {
                 UT_CONTROL_PLANE_DEBUG("DATA RECEIVED\n");
                 call_callback_on_match(msg, pInternal);
+
+                free(msg->message);
+                msg->message = NULL;
                 free(msg);
+                msg = NULL;
             }
             break;
 
             default:
             {
-
+                UT_CONTROL_PLANE_ERROR("Unknown message type received: %d\n", msg->status);
+                if( msg)
+                {
+                    free(msg);
+                    msg = NULL;
+                }
             }
             break;
         }
@@ -329,26 +347,26 @@ static int callback_http(struct lws *wsi, enum lws_callback_reasons reason, void
         case LWS_CALLBACK_HTTP_BODY_COMPLETION:
         {
             UT_CONTROL_PLANE_DEBUG("LWS_CALLBACK_HTTP_BODY_COMPLETION\n");
-            if (perSessionData != NULL)
+            if (!perSessionData)
+                break;
+
+            msg.size = (int)perSessionData->post_data_len;
+            msg.status = DATA_RECIEVED;
+            msg.message = malloc(msg.size + 1);
+            if (!msg.message)
             {
-                UT_CONTROL_PLANE_DEBUG("LWS_CALLBACK_HTTP_BODY_COMPLETION, perSessionData not NULL\n");
-                msg.message = malloc((int)perSessionData->post_data_len + 1);
-                assert(msg.message != NULL);
-                if (msg.message == NULL)
-                {
-                    UT_CONTROL_PLANE_ERROR("Malloc failed\n");
-                    break;
-                }
-                msg.size = (int)perSessionData->post_data_len;
-                msg.status = DATA_RECIEVED;
-                strncpy(msg.message, (const char *)perSessionData->post_data, perSessionData->post_data_len);
-                msg.message[perSessionData->post_data_len] = '\0';
-                // UT_CONTROL_PLANE_DEBUG("Received message:\n %s\n", msg.message);
-                enqueue_message(&msg, pInternal);
-                // char response[] = "{\"status\": \"success\"}";
-                // lws_write(wsi, (unsigned char *)response, strlen(response), LWS_WRITE_HTTP);
-                return 1; // HTTP request handled
+                UT_CONTROL_PLANE_ERROR("Malloc failed\n");
+                break;
             }
+
+            memcpy(msg.message, perSessionData->post_data, msg.size);
+            msg.message[msg.size] = '\0';
+
+            enqueue_message(&msg, pInternal);
+            /* Ownership of msg.message is now transferred to the queue; service_state_machine() frees it. */
+
+            perSessionData->post_data_len = 0; // reset buffer
+            return 1;                          // handled
         }
 
         default:

--- a/src/ut_kvp.c
+++ b/src/ut_kvp.c
@@ -116,22 +116,6 @@ ut_kvp_status_t ut_kvp_open(ut_kvp_instance_t *pInstance, char *fileName)
         return UT_KVP_STATUS_FILE_OPEN_ERROR;
     }
 
-    if(pInternal->fy_handle)
-    {
-        merge_nodes(fy_document_root(pInternal->fy_handle), fy_document_root(fy_document_build_from_file(NULL, fileName)));
-    }
-    else
-    {
-        pInternal->fy_handle = fy_document_create(NULL);
-    }
-
-    if (NULL == pInternal->fy_handle)
-    {
-        UT_LOG_ERROR("Unable to parse file/memory");
-        ut_kvp_close( pInstance );
-        return UT_KVP_STATUS_PARSING_ERROR;
-    }
-
     struct fy_document *srcDoc = fy_document_build_from_file(NULL, fileName);
 
     if(fy_document_resolve(srcDoc) != 0)
@@ -141,12 +125,28 @@ ut_kvp_status_t ut_kvp_open(ut_kvp_instance_t *pInstance, char *fileName)
         return UT_KVP_STATUS_PARSING_ERROR;
     }
 
+    if(pInternal->fy_handle)
+    {
+        merge_nodes(fy_document_root(pInternal->fy_handle), fy_document_root(srcDoc));
+    }
+    else
+    {
+        pInternal->fy_handle = fy_document_create(NULL);
+        if (NULL == pInternal->fy_handle)
+        {
+            UT_LOG_ERROR("Unable to parse file/memory");
+            ut_kvp_close(pInstance);
+            return UT_KVP_STATUS_PARSING_ERROR;
+        }
+    }
+
     node = process_node_copy(fy_document_root(srcDoc), pInternal->fy_handle, 0);
 
     if (node == NULL)
     {
         UT_LOG_ERROR("Unable to process node");
         ut_kvp_close(pInstance);
+        fy_document_destroy(srcDoc);
         return UT_KVP_STATUS_PARSING_ERROR;
     }
 
@@ -175,19 +175,9 @@ ut_kvp_status_t ut_kvp_openMemory(ut_kvp_instance_t *pInstance, char *pData, uin
 
     cData = strdup((const char*)pData);
 
-    if (pInternal->fy_handle)
+    if (cData == NULL)
     {
-        merge_nodes(fy_document_root(pInternal->fy_handle), fy_document_root(fy_document_build_from_malloc_string(NULL, cData, length)));
-    }
-    else
-    {
-        pInternal->fy_handle = fy_document_create(NULL);
-    }
-
-    if (NULL == pInternal->fy_handle)
-    {
-        UT_LOG_ERROR("Unable to parse file/memory");
-        ut_kvp_close( pInstance );
+        UT_LOG_ERROR("Memory allocation failure");
         return UT_KVP_STATUS_PARSING_ERROR;
     }
 
@@ -200,12 +190,29 @@ ut_kvp_status_t ut_kvp_openMemory(ut_kvp_instance_t *pInstance, char *pData, uin
         return UT_KVP_STATUS_PARSING_ERROR;
     }
 
+    if (pInternal->fy_handle)
+    {
+        merge_nodes(fy_document_root(pInternal->fy_handle), fy_document_root(srcDoc));
+    }
+    else
+    {
+        pInternal->fy_handle = fy_document_create(NULL);
+        if (NULL == pInternal->fy_handle)
+        {
+            UT_LOG_ERROR("Unable to parse file/memory");
+            ut_kvp_close(pInstance);
+            fy_document_destroy(srcDoc);
+            return UT_KVP_STATUS_PARSING_ERROR;
+        }
+    }
+
     node = process_node_copy(fy_document_root(srcDoc), pInternal->fy_handle, 0);
 
     if (node == NULL)
     {
         UT_LOG_ERROR("Unable to process node");
         ut_kvp_close(pInstance);
+        fy_document_destroy(srcDoc);
         return UT_KVP_STATUS_PARSING_ERROR;
     }
 
@@ -1010,7 +1017,6 @@ static struct fy_node* process_node_copy(struct fy_node *srcNode, struct fy_docu
 
 static void merge_nodes(struct fy_node *mainNode, struct fy_node *includeNode)
 {
-
     if (mainNode == NULL)
     {
         UT_LOG_ERROR("Main node is invalid");
@@ -1025,8 +1031,24 @@ static void merge_nodes(struct fy_node *mainNode, struct fy_node *includeNode)
 
     if (fy_node_is_scalar(mainNode))
     {
-        fy_node_create_scalar_copy(fy_node_document(mainNode), fy_node_get_scalar(includeNode, NULL), fy_node_get_scalar_length(includeNode));
-        free((void *)fy_node_get_scalar(includeNode, NULL));
+        const char *scalar = fy_node_get_scalar(includeNode, NULL);
+        size_t scalar_len = fy_node_get_scalar_length(includeNode);
+
+        if (scalar)
+        {
+            struct fy_node *new_scalar = fy_node_create_scalar_copy(fy_node_document(mainNode), scalar, scalar_len);
+            if (!new_scalar)
+            {
+                UT_LOG_ERROR("Failed to create scalar copy");
+                return;
+            }
+
+            mainNode = new_scalar;
+        }
+        else
+        {
+            UT_LOG_ERROR("Included scalar is NULL");
+        }
     }
     else if (fy_node_is_mapping(mainNode) && fy_node_is_mapping(includeNode))
     {

--- a/tests/src/ut_test_kvp.c
+++ b/tests/src/ut_test_kvp.c
@@ -150,6 +150,7 @@ void test_ut_kvp_open( void )
     ut_kvp_close(pInstance);
     UT_LOG_STEP("ut_kvp_close(4) - Positive");
     ut_kvp_close(pInstance);
+    ut_kvp_destroyInstance( pInstance );
 }
 
 void test_ut_kvp_open_memory( void )

